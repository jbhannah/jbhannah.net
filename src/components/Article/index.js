import React from 'react'
import { DateTime } from 'luxon'

import styles from './styles.module.css'
import MarginFix from '../MarginFix'

function localDateFromDateTime(date) {
  return DateTime.fromISO(date).setZone('America/Phoenix').toLocaleString(DateTime.DATE_HUGE)
}

const Article = ({ data, list }) => {
  const TitleHeader = list ? 'h2' : 'h1'
  let title

  if (data.frontmatter.link) {
    title = <a href={data.frontmatter.link}>{data.frontmatter.title}</a>
  } else if (list) {
    title = <a href={'/' + data.fields.slug}>{data.frontmatter.title}</a>
  } else {
    title = data.frontmatter.title
  }

  return (
    <article>
      <header>
        <TitleHeader>{title}</TitleHeader>
        <p className={styles.metadata}>
          <time dateTime={data.frontmatter.date}>
            {localDateFromDateTime(data.frontmatter.date)}
          </time>
          {' â€” '}
          {data.timeToRead} minute read
        </p>
        { list && <MarginFix /> }
      </header>
      <section>
        {data.content}
      </section>
    </article>
  )
}

export default Article
