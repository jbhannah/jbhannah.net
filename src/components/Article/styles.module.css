.metadata {
  font-size: 0.75rem;
  margin-bottom: 0.79rem;
}

.footer {
  margin-top: 1.58rem;
}
