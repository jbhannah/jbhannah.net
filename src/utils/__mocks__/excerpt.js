module.exports = jest.fn().mockImplementation(({ htmlAst }) => ({
  ...htmlAst,
  excerpted: true,
}))
