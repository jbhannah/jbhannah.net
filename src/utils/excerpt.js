const unified = require('unified')
const remark = require('remark-parse')
const html = require('remark-html')

function excerpt (file) {
  const output = unified()
    .use(remark, {
      commonmark: true,
      footnotes: true
    })
    .use(html)
    .processSync(file.content)
    .contents

  if (file.data.hasOwnProperty('link')) {
    file.excerpt = output
  } else {
    file.excerpt = output
      .slice(output.indexOf('<p>'), output.indexOf('</p>') + 4)
      .replace(/<sup id="fnref-\S+">.+?<\/sup>/, '')
      .replace(/<a href="\S+">([\S\s]+?)<\/a>/g, '$1')
  }
}

module.exports = excerpt
