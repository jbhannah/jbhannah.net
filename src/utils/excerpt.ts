import remove from "unist-util-remove"
import stripPosition from "unist-util-remove-position"

const excerpt = ({ htmlAst, frontmatter }) => {
  if (frontmatter.link === null) {
    htmlAst.children = [htmlAst.children.filter((n) => n.tagName === "p")[0]]
  }

  stripPosition(htmlAst, true)
  remove(htmlAst, { tagName: "sup" })

  return htmlAst
}

export default excerpt
