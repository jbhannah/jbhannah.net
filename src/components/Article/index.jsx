import { graphql } from "gatsby"
import { DateTime } from "luxon"
import React from "react"
import excerpt from "../../utils/excerpt"
import Heading from "../Heading"
import Link from "../Link"
import MarginFix from "../MarginFix"
import Markdown from "../Markdown"

const localDateFromDateTime = date =>
  DateTime.fromISO(date)
    .setZone("America/Phoenix")
    .toLocaleString(DateTime.DATE_HUGE)

const Article = ({ list, article }) => {
  const { fields, frontmatter, timeToRead } = article
  const content = list ? excerpt(article) : article.htmlAst
  let title

  if (frontmatter.link) {
    title = <Link href={frontmatter.link}>{frontmatter.title}</Link>
  } else if (list) {
    title = <Link href={`/${fields.slug}`}>{frontmatter.title}</Link>
  } else {
    title = frontmatter.title
  }

  return (
    <article>
      <header>
        <Heading level={list ? "h2" : "h1"}>{title}</Heading>
        <p css={{ fontSize: "0.75rem", marginBottom: "0.79rem" }}>
          <time dateTime={frontmatter.date}>
            {localDateFromDateTime(frontmatter.date)}
          </time>
          {" — "}
          {timeToRead} minute read
        </p>
        {list && <MarginFix />}
      </header>
      <Markdown htmlAst={content} />
      {list && (
        <footer css={{ marginTop: "1.58rem" }}>
          <Link href={`/${fields.slug}`}>
            {frontmatter.link ? "Permalink" : "Read More…"}
          </Link>
        </footer>
      )}
    </article>
  )
}

export default Article

export const query = graphql`
  fragment ArticleInformation on MarkdownRemark {
    fields {
      slug
    }
    frontmatter {
      title
      date
      link
    }
    htmlAst
    timeToRead
  }
`
