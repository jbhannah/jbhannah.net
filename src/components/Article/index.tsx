import styled, { CSSObject } from "@emotion/styled"
import { graphql } from "gatsby"
import { DateTime } from "luxon"
import * as React from "react"
import excerpt from "../../utils/excerpt"
import { rhythm } from "../../utils/typography"
import { Heading as _Heading } from "../Heading"
import { Link } from "../Link"
import { MarginFix } from "../MarginFix"
import { Markdown } from "../Markdown"
import { ArticleInformation } from "./__generated__/ArticleInformation"

const localDateFromDateTime = (date: string) =>
  DateTime.fromISO(date)
    .setZone("America/Phoenix")
    .toLocaleString(DateTime.DATE_HUGE)

const headingCSS: CSSObject = { marginBottom: rhythm(1 / 2) }

const Heading = styled(_Heading)(headingCSS)
const Metadata = styled.p({ fontSize: rhythm(1 / 2), ...headingCSS })
const Footer = styled.footer({ marginTop: rhythm(1) })

interface ArticleProps {
  list?: boolean
  article: ArticleInformation
}

export const Article: React.FunctionComponent<ArticleProps> = ({
  list,
  article,
}) => {
  const { fields, frontmatter, timeToRead } = article
  const content = list ? excerpt(article) : article.htmlAst
  let title

  if (frontmatter.link) {
    title = <Link href={frontmatter.link}>{fields.title}</Link>
  } else if (list) {
    title = <Link href={`${fields.slug}`}>{fields.title}</Link>
  } else {
    title = fields.title
  }

  return (
    <article>
      <header>
        <Heading level={list ? "h2" : "h1"}>{title}</Heading>
        <Metadata>
          <time dateTime={frontmatter.date}>
            {localDateFromDateTime(frontmatter.date)}
          </time>
          {" — "}
          {timeToRead} minute read
        </Metadata>
        {list && <MarginFix />}
      </header>
      <Markdown htmlAst={content} />
      {list && (
        <Footer>
          <Link href={`${fields.slug}`}>
            {frontmatter.link ? "Permalink" : "Read More…"}
          </Link>
        </Footer>
      )}
    </article>
  )
}

export const query = graphql`
  fragment ArticleInformation on MarkdownRemark {
    fields {
      slug
      title
    }
    frontmatter {
      date
      link
    }
    excerpt
    htmlAst
    timeToRead
  }
`
