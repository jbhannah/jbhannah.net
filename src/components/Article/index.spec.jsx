import { shallow } from "enzyme"
import React from "react"
import Article from "."
import excerpt from "../../utils/excerpt"

jest.mock("../../utils/excerpt")

const article = {
  fields: {
    slug: "/test-slug",
    title: "Test Article",
  },
  frontmatter: {
    date: "2019-02-05T01:17:07.551Z",
  },
  htmlAst: {
    excerpted: false,
  },
  timeToRead: 5,
}

const site = {
  siteMetadata: {
    title: "Test Site",
  },
}

const linkArticle = Object.assign({}, article, {
  frontmatter: { link: "https://example.com/test-link" },
})

const TestArticle = (props) => shallow(<Article {...props} />)

describe("Article", () => {
  describe("standalone", () => {
    const tree = TestArticle({ article, site })
    expect(excerpt).not.toBeCalled()

    it("renders correctly", () => {
      expect(tree).toMatchSnapshot()
    })

    it("displays the article title in an h1", () => {
      const heading = tree.find("header > Heading")
      expect(heading.prop("level")).toBe("h1")
      expect(heading.render().text()).toBe("Test Article")
    })

    it("does not display the article title as a link", () => {
      const heading = tree.find("header > Heading")
      expect(heading.children("Link")).toHaveLength(0)
    })

    it("displays the publication date", () => {
      const time = tree.find("time")
      expect(time.prop("dateTime")).toEqual("2019-02-05T01:17:07.551Z")
      expect(time.text()).toEqual("Monday, February 4, 2019")
    })

    it("renders the non-excerpted body", () => {
      expect(tree.find("Markdown").prop("htmlAst")).toHaveProperty(
        "excerpted",
        false
      )
    })

    it("does not render a footer", () => {
      expect(tree.find("footer")).toHaveLength(0)
    })

    describe("with a link", () => {
      const tree = TestArticle({ article: linkArticle, site })

      it("displays the article title as a link", () => {
        const title = tree.find("header > Heading > Link")
        expect(title.prop("href")).toBe("https://example.com/test-link")
      })
    })
  })

  describe("in a list", () => {
    const tree = TestArticle({ article, site, list: true })
    expect(excerpt).toBeCalled()

    it("renders correctly", () => {
      expect(tree).toMatchSnapshot()
    })

    it("displays the article title in an h2", () => {
      const heading = tree.find("header > Heading")
      expect(heading.prop("level")).toBe("h2")
      expect(heading.render().text()).toBe("Test Article")
    })

    it("displays the article title as a link", () => {
      const title = tree.find("header > Heading > Link")
      expect(title.prop("href")).toBe("/test-slug")
      expect(title.render().text()).toBe("Test Article")
    })

    it("renders the excerpted body", () => {
      expect(tree.find("Markdown").prop("htmlAst")).toHaveProperty(
        "excerpted",
        true
      )
    })

    it('renders a footer with "Read More…" link', () => {
      const footer = tree.find("footer")
      expect(footer.find("Link").prop("href")).toBe("/test-slug")
      expect(footer.render().text()).toBe("Read More…")
    })

    describe("with a link", () => {
      const tree = TestArticle({ article: linkArticle, site, list: true })

      it("displays the article title as a link", () => {
        const title = tree.find("header > Heading > Link")
        expect(title.prop("href")).toBe("https://example.com/test-link")
      })

      it('renders a footer with "Permalink" link', () => {
        const footer = tree.find("footer")
        expect(footer.find("Link").prop("href")).toBe("/test-slug")
        expect(footer.render().text()).toBe("Permalink")
      })
    })
  })
})
