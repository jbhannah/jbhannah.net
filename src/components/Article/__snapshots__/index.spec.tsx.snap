// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article in a list renders correctly 1`] = `
<article>
  <header>
    <Styled(Heading)
      level="h2"
    >
      <Link
        href="/test-slug"
      >
        Test Article
      </Link>
    </Styled(Heading)>
    <Styled(p)>
      <time
        dateTime="2019-02-05T01:17:07.551Z"
      >
        Monday, February 4, 2019
      </time>
       — 
      5
       minute read
    </Styled(p)>
    <Styled(span) />
  </header>
  <Markdown
    htmlAst={
      Object {
        "excerpted": true,
      }
    }
  />
  <Styled(footer)>
    <Link
      href="/test-slug"
    >
      Read More…
    </Link>
  </Styled(footer)>
</article>
`;

exports[`Article standalone renders correctly 1`] = `
<article>
  <header>
    <Styled(Heading)
      level="h1"
    >
      Test Article
    </Styled(Heading)>
    <Styled(p)>
      <time
        dateTime="2019-02-05T01:17:07.551Z"
      >
        Monday, February 4, 2019
      </time>
       — 
      5
       minute read
    </Styled(p)>
  </header>
  <Markdown
    htmlAst={
      Object {
        "excerpted": false,
      }
    }
  />
</article>
`;
